contract/src/offer-up.contract.js

```diff
@@ -1,70 +1,143 @@
-/** @file Contract to mint and sell Item NFTs. */
+/**
+ * @file Contract to mint and sell a few Item NFTs at a time.
+ *
+ * We declare variables (including functions) before using them,
+ * so you may want to skip ahead and come back to some details.
+ * @see {start} for the main contract entrypoint
+ *
+ * @see {@link https://docs.agoric.com/guides/zoe/|Zoe Overview} for a walk-thru of this contract
+ * @see {@link https://docs.agoric.com/guides/js-programming/hardened-js.html|Hardened JavaScript}
+ * for background on `harden` and `assert`.
+ */
 // @ts-check
 
 import { Far } from '@endo/far';
 import { M, getCopyBagEntries } from '@endo/patterns';
 import { AmountMath, AssetKind } from '@agoric/ertp/src/amountMath.js';
 import { AmountShape } from '@agoric/ertp/src/typeGuards.js';
-// Use the deprecated atomicRearrange API
-// for compatibility with mainnet1B.
+import { atomicRearrange } from '@agoric/zoe/src/contractSupport/atomicTransfer.js';
 import '@agoric/zoe/exported.js';
 
-import { makeTracer } from './debug.js';
-
 const { Fail, quote: q } = assert;
 
-const trace = makeTracer('OfferUp', true);
+// #region bag utilities
+/** @type { (xs: bigint[]) => bigint } */
+const sum = xs => xs.reduce((acc, x) => acc + x, 0n);
 
-/** @param {Amount<'copyBag'>} amt */
-const bagValueSize = amt => {
-  /** @type {[unknown, bigint][]} */
-  const entries = getCopyBagEntries(amt.value); // XXX getCopyBagEntries returns any???
-  const total = entries.reduce((acc, [_, qty]) => acc + qty, 0n);
-  return total;
+/**
+ * @param {import('@endo/patterns').CopyBag} bag
+ * @returns {bigint[]}
+ */
+const bagCounts = bag => {
+  const entries = getCopyBagEntries(bag);
+  return entries.map(([_k, ct]) => ct);
 };
+// #endregion
 
 /**
- * @param {ZCF<{tradePrice: Amount}>} zcf
+ * This contract is parameterized by terms for price and,
+ * optionally, a maximum number of items sold for that price (default: 3).
+ *
+ * @typedef {{
+ *   tradePrice: Amount;
+ *   maxItems?: bigint;
+ * }} ItemTerms
+ */
+
+export const meta = {
+  customTermsShape: M.splitRecord(
+    { tradePrice: AmountShape },
+    { maxItems: M.bigint() },
+  ),
+};
+// compatibility with an earlier contract metadata API
+export const customTermsShape = meta.customTermsShape;
+
+/**
+ * Start a contract that
+ *   - creates a new non-fungible asset type for Items, and
+ *   - handles offers to buy up to `maxItems` items at a time.
+ *
+ * As is typical in Zoe contracts, the flow is:
+ *   1. contract does internal setup
```

Comment:
```suggestions
 *   1. contract does internal setup (mint assets)
 ```
 
Can we expand on 'internal setup'?

Reply:
Well, it doesn't mint any assets at startup. It creates a mint, but doesn't use it until an offer is made.

The point of this comment was to clarify the order that readers should use to look at the code. They can see the rest for themselves, I would think.

In fact, now that I think about it, this part of the comment doesn't belong on the doctring of `start`; it's not aimed at clients of `start`. It's an internal comment aimed at readers of the implementation. I'm moving it to the file comment.